Refactor: Change time variable type from int64_t to std::chrono::seconds in net_processing.cpp

shaavan opened the pull request:
- This is a follow-up to PR #23758
- This changes the remaining time variable in `net_processing.cpp` from **int64_t** to **std::chrono::seconds**

shaavan committed:
Change time variable type to std::chrono::seconds in src/net_processing.cpp

- This commit is a followup to commit: 60b579
- This changes the remaining time variable in net_processing.cpp from
int64_t to std::chrono

shaavan commented:
A gentle ping, @MarcoFalke

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -58,9 +58,9 @@ static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4;
 static constexpr int64_t CHAIN_SYNC_TIMEOUT = 20 * 60; // 20 minutes
 /** How frequently to check for stale tips, in seconds */
-static constexpr int64_t STALE_CHECK_INTERVAL = 10 * 60; // 10 minutes
+static constexpr std::chrono::seconds STALE_CHECK_INTERVAL{10 * 60}; // 10 minutes
```
```suggestion
static constexpr std::chrono::minutes STALE_CHECK_INTERVAL{10};
```

nit. Also:

```
The subject line of commit hash f8b810b8e15b72f5cff4c6e27bea3f6a64fd0c3d is followed by a non-empty line. Subject lines should always be followed by a blank line.

shaavan replied:
Addressed in 8e0ba04b2e6e9ecb11b541ecc361c3a9fe5e9fde

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -58,9 +58,9 @@ static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4;
 /** How frequently to check for stale tips, in seconds */
-static constexpr int64_t STALE_CHECK_INTERVAL = 10 * 60; // 10 minutes
+static constexpr std::chrono::seconds STALE_CHECK_INTERVAL{10 * 60}; // 10 minutes
 /** How frequently to check for extra outbound peers and disconnect, in seconds */
```
```suggestion
/** How frequently to check for extra outbound peers and disconnect */
```

nit

shaavan replied:
Addressed in 8e0ba04

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4335,15 +4335,15 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
         // outbound peer
         if (!fImporting && !fReindex && m_connman.GetNetworkActive() && m_connman.GetUseAddrmanOutgoing() && TipMayBeStale()) {
-            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", time_in_seconds - m_last_tip_update);
+            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", (time_in_seconds - m_last_tip_update).count());
```
```suggestion
            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", count_seconds(time_in_seconds - m_last_tip_update));
```

nit: Could clarify the unit

shaavan replied:
Addressed in 8e0ba04

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -947,10 +947,10 @@ bool PeerManagerImpl::TipMayBeStale()
     const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
-    if (m_last_tip_update == 0) {
-        m_last_tip_update = GetTime();
+    if (m_last_tip_update == std::chrono::seconds::zero()) {
```
```suggestion
    if (m_last_tip_update.count() == 0) {
```

I'd use this, but there is really no difference, so both are fine.

shaavan replied:
Better to keep the coding usage throughout the codebase consistent.

Addressed in 8e0ba04

maflcko reviewed (approved):
ack. left some nits

shaavan commented:
Updated from f8b810b8e15b72f5cff4c6e27bea3f6a64fd0c3d to 8e0ba04b2e6e9ecb11b541ecc361c3a9fe5e9fde ([pr23801.01](https://github.com/shaavan/bitcoin/tree/pr23801.01) -> [pr23801.02](https://github.com/shaavan/bitcoin/tree/pr23801.02))

Changes
- Addressed nits, suggested by @MarcoFalke

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -58,9 +58,9 @@ static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4;
 /** Timeout for (unprotected) outbound peers to sync to our chainwork, in seconds */
 static constexpr int64_t CHAIN_SYNC_TIMEOUT = 20 * 60; // 20 minutes
 /** How frequently to check for stale tips, in seconds */
```
```suggestion
/** How frequently to check for stale tips */
```

shaavan replied:
Addressed in 1a1fafa4fdcbe7b0d1ef8df466a71bd3e65195d3

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -541,7 +541,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** When our tip was last updated. */
-    std::atomic<int64_t> m_last_tip_update{0};
+    std::chrono::seconds m_last_tip_update{0};
```
is atomic no longer needed?

shaavan replied:
Initially, I removed it by mistake. But when I tried to initialize var as
```
std::atomic<std::chrono::seconds> m_last_tip_update{0};
```
It is causing a compilation error.

I tried other ways to initialize it, but they don't seem to work.  I searched for a solution on the internet, and I came around this bug report which is in a very similar spirit to this problem: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54025

So how do you suggest I should proceed now:
- Revert this variable type to `std::atomic<int64_t>`
- Keep the current implementation, which the `std::atomic`

laanwj replied:
`atomic` works for specific concrete integer types, it doesn't work for arbitrary data structures, you'll have to use a full mutex. I don't think this is worth it here so might as well stick with a `int64_t`.

MarcoFalke replied:
chronos are plain integers (they only have a bit of type safety added at compile time).

```cpp
atomic<seconds> a{0s};
```

should compile, like in all other places where `atomic<seconds>` is used.

shaavan replied:
`atomic<seconds> a{0}` was not working but for some reason `atomic<seconds> a{0s}` worked.

Addressed in 1a1fafa4fdcbe7b0d1ef8df466a71bd3e65195d3

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4335,15 +4335,15 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
     LOCK(cs_main);
 
-    int64_t time_in_seconds = GetTime();
+    auto time_in_seconds{GetTime<std::chrono::seconds>()};
```
Maybe rename to `now` for brevity?

shaavan replied:
Makes sense. I am going to address it in the next update of PR.

shaavan replied:
Addressed in 1a1fafa4fdcbe7b0d1ef8df466a71bd3e65195d3

maflcko reviewed (approved):
more nits

w0xlt reviewed (approved):
crACK 8e0ba04

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
```
#include <chrono>
```
?

shaavan commented:
Updated from 8e0ba04b2e6e9ecb11b541ecc361c3a9fe5e9fde to 1a1fafa4fdcbe7b0d1ef8df466a71bd3e65195d3 ([pr23801.02](https://github.com/shaavan/bitcoin/tree/pr23801.02) -> [pr23801.03](https://github.com/shaavan/bitcoin/tree/pr23801.03))

Addressed comment by @MarcoFalke

Changes:
- Addressed one nit. Removed unnecessary “in seconds.”
- Changed **m_last_tip_update** time from `std::chrono::seconds` -> `std::atomic<std::chrono::seconds>`. Made other related changes, accordingly.
- Renamed **time_in_seconds** to **now** under `void PeerManagerImpl::CheckForStaleTipAndEvictPeers()` definition.

@Hebasto. I think including `<chrono>` in this file would be redundant, as the header `banman.h`, which this file imports, also imports `chrono` in it.

hebasto commented:
@shaavan 
> @hebasto. I think including `<chrono>` in this file would be redundant, as the header `banman.h`, which this file imports, also imports `chrono` in it.

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization:
> Every `.cpp` and `.h` file should `#include` every header file it directly uses classes, functions or other definitions from, even if those headers are already included indirectly through other headers.
>  - _Rationale_: Excluding headers because they are already indirectly included results in compilation failures when those indirect dependencies change. Furthermore, it obscures what the real code dependencies are.

shaavan commented:
Updated from 1a1fafa4fdcbe7b0d1ef8df466a71bd3e65195d3 to d5207c2b75c33a720879ed1e43f3db1156cdcce8
Addressed @hebasto comment.

Changes:
- Added `<chrono>` in the included header as it is a direct dependency in this file.

Thanks, @hebasto, for enlightening me. I will take care of this in the future.

stratospher reviewed (commented):
ACK d5207c2. 

nit: On the same note, wouldn't `#include <atomic>` be needed too?

hebasto commented on src/net_processing.cpp:
```diff
@@ -422,7 +423,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Next time to check for stale tip */
-    int64_t m_stale_tip_check_time{0};
+    std::chrono::seconds m_stale_tip_check_time{0s};
```
While using `std::literals::chrono_literals` why not using `auto` with universal initialization:
```suggestion
    auto m_stale_tip_check_time{0s};
```

It looks concise and expressive, no?
Also could be applied in other places above.

shaavan replied:
That’s a good suggestion. I have used the auto keyword at all the places where it could be used. See 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469
However, I could not use `auto` with the particular example you had taken. Doing so would cause the following error:

```
error: non-static data member declared with placeholder ‘auto’
```

The rationale for this behavior is discussed [here](https://stackoverflow.com/a/11303175).

hebasto reviewed (approved):
ACK d5207c2b75c33a720879ed1e43f3db1156cdcce8

shaavan commented:
Updated from  d5207c2b75c33a720879ed1e43f3db1156cdcce8 to 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469 ([pr23801.03](https://github.com/shaavan/bitcoin/tree/pr23801.03) -> [pr23801.04](https://github.com/shaavan/bitcoin/tree/pr23801.04))
Addressed @Stratospher and @Hebasto comments.

Changes:
- Added atomic in the included header, as it is a direct dependency.
- Used the universal initializer auto, wherever chrono_literals were used.

naumenkogs commented:
ACK 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469

hebasto reviewed (approved):
re-ACK 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469

hebasto commented:
I've restarted the "macOS 12 native [gui, system sqlite only] [no depends]" Cirrus CI task.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -57,10 +59,10 @@ static constexpr auto HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER = 1ms;
 static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4;
 /** Timeout for (unprotected) outbound peers to sync to our chainwork, in seconds */
 static constexpr int64_t CHAIN_SYNC_TIMEOUT = 20 * 60; // 20 minutes
```
I think this can also be switched to chrono? It would probably be the last occurrence and conclude the chrono-refactors?

shaavan replied:
Let's take it in a follow-up!

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -947,10 +949,10 @@ bool PeerManagerImpl::TipMayBeStale()
+        m_last_tip_update = GetTime<std::chrono::seconds>();
     }
-    return m_last_tip_update < GetTime() - consensusParams.nPowTargetSpacing * 3 && mapBlocksInFlight.empty();
+    return count_seconds(m_last_tip_update) < GetTime() - consensusParams.nPowTargetSpacing * 3 && mapBlocksInFlight.empty();
```
I think it would have been better to use type-safe time here (the goal of this pull), because:

* GetTime should eventually be removed
* count_seconds should only be used to "serialize" time values (for example in RPC results or P2P messages)

(see comments in `src/util/time.h`)

shaavan replied:
> I think it would have been better to use the type-safe time here

My first approach was to do precisely this. But since **m_last_tip_update** was `atomic<seconds>` type.
There were two issues I faced:
1. Converting RHS to match the type
2. For some reason, subtraction between atomic types is not possible

I tried to find some way to convert `atomic<seconds>` to `<seconds>`, but I could not find a satisfactory result. That's why I went with the current implementation.

MarcoFalke replied:
I think `m_last_tip_update.load()` will load the value wrapped by the atomic

shaavan replied:
> I think m_last_tip_update.load() will load the value wrapped by the atomic

It worked! Let me open PR with all the suggested changes, and wrap up this chain of PR of going from `int` -> `chrono` as soon as possible. 

Thanks a lot!

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4335,20 +4337,20 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
         // outbound peer
         if (!fImporting && !fReindex && m_connman.GetNetworkActive() && m_connman.GetUseAddrmanOutgoing() && TipMayBeStale()) {
-            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", time_in_seconds - m_last_tip_update);
+            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds ago)\n", count_seconds(now) - count_seconds(m_last_tip_update));
```
nit: I think it might have been better to do the calculation type-safe and then only "serialize" the result of the addition. This would also result in less code.

shaavan replied:
One doubt:

Is there a way to go from `atomic<seconds>` to `<seconds>`
Or do subtraction between two `atomic<seconds>` type variables?

Thanks.

shaavan replied:
>  subtraction between two atomic<seconds> type variables?

For reference, doing so would cause the following error:
```
no match for ‘operator-’ (operand types are ‘std::atomic<std::chrono::duration<long int> >’ and ‘std::atomic<std::chrono::duration<long int> >’)
```

cc @MarcoFalke

MarcoFalke commented:
re-ack. Left two nits and a follow-up idea.
